src/compaction_filter.h

```diff
@@ -0,0 +1,115 @@
+#pragma once
+
+#include "db_impl.h"
+#include "rocksdb/compaction_filter.h"
+
+#include <utility>
+
+namespace rocksdb {
+namespace titandb {
+
+class TitanCompactionFilter final : public CompactionFilter {
+public:
+  explicit TitanCompactionFilter(const TitanDBImpl *db,
+                                 const CompactionFilter *origin, uint32_t cf_id)
+      : titan_db_impl_(db), origin_filter_(origin), column_family_id_(cf_id) {}
+
+  const char *Name() const override { return "TitanCompactionFilter"; }
+
+  Decision FilterV2(int level, const Slice &key, ValueType value_type,
+                    const Slice &value, std::string *new_value,
+                    std::string *skip_until) const override {
+    if (value_type != kBlobIndex) {
+      return origin_filter_->FilterV2(level, key, value_type, value, new_value,
+                                      skip_until);
+    }
+
+    BlobIndex blob_index;
+    Slice origin_value(value.data());
+    Status s = blob_index.DecodeFrom(&origin_value);
+    if (!s.ok()) {
+      // Unable to decode blob index. Keeping the value.
+      return Decision::kKeep;
+    }
+    if (BlobIndex::IsDeletionMarker(blob_index)) {
+      return Decision::kRemove;
+    }
+
+    BlobRecord record;
+    PinnableSlice buffer;
+
+    titan_db_impl_->mutex_.Lock();
+    auto storage =
+        titan_db_impl_->blob_file_set_->GetBlobStorage(column_family_id_)
```

Comment:
hmm, shall we cache `blob_storage` for each compaction filter instance? To avoid the need to lock global mutex for each key. The CF won't be dropped while compaction is still running.

Reply:
set `blob_storage` for each compaction filter instance in ctor.